ompi/util/count_disp_array.h

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2024   Triad National Security, LLC. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+#ifndef OMPI_UTIL_COUNT_DISP_ARRAY_H
+#define OMPI_UTIL_COUNT_DISP_ARRAY_H
+
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdint.h>
+
+enum ompi_count_array_type {
+    OMPI_COUNT_ARRAY_TYPE_INT,
+    OMPI_COUNT_ARRAY_TYPE_SIZE_T,
+};
+
+typedef struct ompi_count_array {
+    enum ompi_count_array_type type;
+    union {
+        const int *int_array;
+        const size_t *size_t_array;
+    } data;
+} ompi_count_array;
+
+/* Initialize an int variant of the count array */
+static inline void ompi_count_array_init(ompi_count_array *array, const int *data)
+{
+    array->type = OMPI_COUNT_ARRAY_TYPE_INT;
+    array->data.int_array = data;
+}
+
+/* Initialize a bigcount variant of the count array */
+static inline void ompi_count_array_init_c(ompi_count_array *array, const size_t *data)
+{
+    array->type = OMPI_COUNT_ARRAY_TYPE_SIZE_T;
+    array->data.size_t_array = data;
+}
+
+#if OPAL_C_HAVE__GENERIC
+#define OMPI_COUNT_ARRAY_INIT(array, data) _Generic((data), \
+                                                    int *: ompi_count_array_init, \
+                                                    const int *: ompi_count_array_init, \
+                                                    size_t *: ompi_count_array_init_c, \
+                                                    const size_t *: ompi_count_array_init_c)(array, data)
+#else
+#define OMPI_COUNT_ARRAY_INIT(array, data) \
+    do { \
+        if (sizeof(*data) == sizeof(int)) { \
+            ompi_count_array_init(array, (const int *) data); \
+        } else if (sizeof(*data) == sizeof(size_t)) { \
+            ompi_count_array_init_c(array, (const size_t *) data); \
+        } \
+    } while (0)
+#endif
+
+/* Return if the internal type is 64-bit or not */
+static inline bool ompi_count_array_is_64bit(ompi_count_array *array)
+{
+    return array->type == OMPI_COUNT_ARRAY_TYPE_SIZE_T;
+}
```

Comment:
`size_t` is no always 64bit. If we revive 32bit support this would technically be wrong so I suggest renaming:

```suggestion
static inline bool ompi_count_array_is_size_t(ompi_count_array *array)
{
    return array->type == OMPI_COUNT_ARRAY_TYPE_SIZE_T;
}
```

Reply:
Hmm, I wonder what to do for the ucc component then. See `coll_ucc_scatterv.c` for example. Basically I cast the inner pointer to `ucc_count_t *` or `ucc_aint_t *`, set it on the arg struct, and then set a flag indicating 64-bit. I guess I could check if `size_t`/`ptrdiff_t` are 64-bit directly in that code? Or maybe this should be using `uint64_t` instead of `size_t`?

Reply:
Mhh, maybe the code there should include a check for `size_t`:

```C
    uint64_t flags = 0;
    if (sizeof(size_t) == 8 && ompi_count_array_is_size_t(scounts)) {
        flags |= UCC_COLL_ARGS_FLAG_COUNT_64BIT;
    }
    if (sizeof(size_t) == 8 && ompi_disp_array_is_size_t(disps)) {
        flags |= UCC_COLL_ARGS_FLAG_DISPLACEMENTS_64BIT;
    }
```

Reply:
I'm thinking this might work:
```c
static inline bool ompi_count_array_is_64bit(ompi_count_array_t *array)
{
    return array->type == OMPI_COUNT_ARRAY_TYPE_SIZE_T && sizeof(size_t) == 8;
}
```